src/codegen/machine_ir.h

```diff
@@ -0,0 +1,180 @@
+
+#include <codegen/codegen_forward.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <vector.h>
+#include <utils.h>
+
+#define DEFINE_MIR_INSTRUCTION_TYPE(type, ...) CAT(MIR_, type),
+typedef enum MIROpcodeCommon {
+  ALL_IR_INSTRUCTION_TYPES(DEFINE_MIR_INSTRUCTION_TYPE)
+  /// Marks beginning of block
+  // TODO: Do we need this?
+  MIR_BLOCK,
+  MIR_COUNT,
+  MIR_ARCH_START = 0x420
+} MIROpcodeCommon;
+#undef DEFINE_MIR_INSTRUCTION_TYPE
+
+typedef enum MIROperandKind {
+  MIR_OP_NONE,
+  MIR_OP_REGISTER,
+  MIR_OP_IMMEDIATE,
+  MIR_OP_BLOCK,
+  MIR_OP_FUNCTION,
+  MIR_OP_NAME,
+  MIR_OP_COUNT
+} MIROperandKind;
+
+typedef struct MIROperandRegister {
+  uint32_t value;
+  uint16_t size;
+} MIROperandRegister;
+typedef const char* MIROperandName;
+typedef int64_t MIROperandImmediate;
+typedef IRBlock* MIROperandBlock;
+typedef IRFunction* MIROperandFunction;
+
+typedef unsigned int MIRRegister;
+
+typedef struct MIRValue_x86_64 {
+  uint8_t instruction_form;
+  uint16_t instruction;
+  uint8_t reg_src_sz;
+  uint8_t reg_src;
+  uint8_t reg_dst_sz;
+  uint8_t reg_dst;
+  uint8_t reg_addr;
+  int64_t immediate;
+  int64_t offset;
+  const char *name;
+  IRBlock *ir_block;
+  IRFunction *ir_function;
+} MIRValue_x86_64;
+
+typedef struct MIROperand {
+  MIROperandKind kind;
+  union {
+    MIROperandRegister reg;
+    MIROperandImmediate imm;
+    MIROperandName name;
+    MIROperandBlock block;
+    MIROperandFunction function;
+  } value;
+} MIROperand;
+
+#define MIR_OPERAND_SSO_THRESHOLD 3
+
+// NOTE: There is no way of keeping track of which ISA a particular
+// instruction is for; that's up to each backend that uses it!
+typedef struct MIRInstruction {
+  size_t id;
+
+  MIRRegister reg;
+
+  /// An MIR opcode is some integer that has the 1:1 IR types from 0->N
+  /// and then after that each ISA starts it's own opcode enum for
+  /// architecture-specific instructions.
+  /// The idea is that each ISA gets fed an MIR full of regular IR
+  /// instructions and it's job is to translate them into (or build a
+  /// new MIR with) the arch-specific opcodes.
+  uint32_t opcode;
+
+  uint8_t operand_count;
+  union {
+    MIROperand arr[MIR_OPERAND_SSO_THRESHOLD];
+    Vector(MIROperand) vec;
+  } operands;
+
+  // Architecture-specific values.
+  union {
+    MIRValue_x86_64 x64;
+  };
+
+  MIRBlock *block;
+
+  // Keep track of originating IR instruction.
+  IRInstruction *origin;
+
+  // If an architecture lowers a generically-lowered MIR instruction,
+  // store the arch-lowered MIR instruction in the generic instruction
+  // so that we may properly update references going forward.
+  struct MIRInstruction *lowered;
+
+} MIRInstruction;
+
+typedef struct MIRBlock {
+  string name;
+
+  // Index within `instructions` of MIRFunction pointing to first
+  // instruction executed in basic block.
+  MIRRegister entry;
+  // Index within `instructions` of MIRFunction pointing to last
+  // instruction of basic block.
+  MIRRegister exit;
```

Comment:
I can see this becoming a problem if one MIR instruction is replaced with several instructions or if multiple instructions are folded into one, as is a common operation during instruction selection.

Reply:
Interesting; in my head I was thinking that by the time we get to MIR we wouldn't really be doing much replacing of instructions or anything, as that would all happen in the IR or during the pattern matching and creation of the lowered MIR from the general MIR. If we do end up having to do a lot of the same operations on the MIR as on the IR (like replacing instructions, keeping track of uses, etc), then yeah this isn't a good move. I just figured MIR was pretty much static as everything happened in IR (like lowering) or in genMIR->loweredMIR stage.

Reply:
Well, you still need to do instruction selection on the MIR. I’d say we either do blocks just like in the IR or abolish blocks completely, but I’m not sure this approach here would work too well...

Reply:
The instruction selection happens during the genMIR -> loweredMIR stage; this stage will eventually be declarative through the pattern table, where patterns of generic MIR instructions will be used to add new loweredMIR instructions to a new set of loweredMIR. I see no reason to mutate the generic MIR into lowered MIR when we can use it as an immutable input to possibly generate multiple sets of lowered MIR, if a user may want to compile for multiple arch/target combos

Reply:
Okay I redact my statement we should absolutely mutate the MIR because of virtual register numbering and reference invalidation.